public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts

```diff
@@ -137,33 +141,42 @@ export const queriesAndExpressionsReducer = createReducer(initialState, (builder
       state.queries = [...state.queries, ...payload];
     })
     .addCase(updateExpression, (state, { payload }) => {
-      state.queries = state.queries.map((query) => {
-        const dataSourceAlertQuery = findDataSourceFromExpression(state.queries, payload.expression);
+      const queryToUpdate = state.queries.find((query) => query.refId === payload.refId);
+
+      if (!queryToUpdate) {
+        return;
+      }
+
+      queryToUpdate.model = payload;
+
+      if (payload.type === ExpressionQueryType.resample && payload.expression) {
```

Comment:
nit: can we add a comment here explaining why we do what we do here for resample expressions?